### PR TITLE
`Programming exercises`: Fix an issue in which the submission policy was not enforced

### PR DESCRIPTION
**Pull request by frederik-kt**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
The lock repository submission policy is currently not enforced.

### Description
This PR fixes the retrieval of the submission policy from the database. This means that the online editor in the client will show an error if a student pushes to their repository after they reached the submission limit. The correct locking of the client will be fixed in a follow-up PR.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise with a lock repository submission policy configured

1. Log in to Artemis as the student and go to the exercise with the submission policy.
2. Submit until you reached the submission limit (either via the online editor or via the local Git client).
3. Make sure you cannot push after the submission limit is reached.
4. Try to push to the student's repository as the instructor and make sure you are still able to do that.

#### Exam Mode Testing
- Repeat the steps from above for an exam exercise.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [RepositoryProgrammingExerciseParticipationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4821-3/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.repository/RepositoryProgrammingExerciseParticipationResource.html) | 80% | ok |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java [hunk 1]
```diff
@@ -86,9 +86,7 @@ Repository getRepository(Long participationId, RepositoryActionType repositoryAc
         }
 
         // Add submission policy to the programming exercise.
-        if (programmingExercise.getSubmissionPolicy() == null) {
```
Comment (frederik-kt) [inline src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java:89]: The way the participation and programming exercise are retrieved above, the submission policy will always be "not initialized" (not null!).

**Review by simon-hng — APPROVED** [review]
Sets submission policy correctly for `lock repository`

**Review by tobias-lippert — COMMENTED** [review]
submission policy limit seems to be enforced but the first push that should be forbidden showed no graded result and the submission counter wasn't increased correctly.

Reply: The broken submission counter is not related to this PR. If no result was created, something went wrong in the continuous integration server which is also not related to this PR.

**Review by sven0311 — COMMENTED** [review]
image

When spamming the submit button of the OnlineIDE I was able to get 3 results but the submission policy is 2. Also the submission count shows only 1/2 . 

Reply: It is unevitable currently to have these submissions slip through because submissions are only counted toward the submission limit if there is a result present for them. We accept this as this only happens if the student submits in rapid succession and this would not lead to them gaining an advantage. The broken submission count is unrelated to this PR.

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionPolicyService.java [hunk 2]
```diff
@@ -319,6 +319,13 @@ private int getParticipationSubmissionCount(Participation participation) {
         if (submissions != null && !submissions.isEmpty()) {
             submissionCompensation = submissions.iterator().next().getResults().isEmpty() ? 1 : 0;
         }
+        // Note: The way the participation submissions are counted here (filtering out submissions without results), leads to unexpected behavior when the user submits to their
+        // repository in rapid succession.
+        // When the user submits while the result for the previous submission is not yet available, the previous submission will not be counted.
+        // This means that the user is able to submit more often than the allowed number of submissions, if a lock repository policy is configured.
+        // As these submissions have to happen in quick succession, this does not constitute an advantage for the student and the behaviour is acceptable for now.
+        // Nonetheless, this method should be reassessed to determine whether it is also feasible to count the submissions without taking results into account.
+        // TODO: Reassess method to determine whether it is also feasible to count the submissions without taking results into account.
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/SubmissionPolicyService.java:328]: This is the intended behavior. Students should only loose a submission, when they got an advantage i.e. a result from the build system

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/SubmissionPolicyService.java (hunk 2 above)
Comment (frederik-kt) [reply src/main/java/de/tum/in/www1/artemis/service/SubmissionPolicyService.java]: Removed the TODO

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts2. Submission limit was enforced.
Code also LGTM

**Review by JohannesStoehr — APPROVED** [review]
Code changes and explanation about the submission policy look good to me

**Pull request merged**